receiver/snmpreceiver/scraper.go

```diff
@@ -16,19 +16,88 @@ package snmpreceiver // import "github.com/open-telemetry/opentelemetry-collecto
+	errMsgBadValueType                          = `returned metric SNMP data type for OID '%s' is not supported`
+	errMsgIndexedBadValueType                   = `returned metric SNMP data type for OID '%s' from column OID '%s' is not supported`
+	errMsgIndexedAttributesBadValueType         = `returned attribute SNMP data type for OID '%s' from column OID '%s' is not supported`
+	errMsgAttributeEmptyValue                   = `metric OID attribute value is blank`
+	errMsgOIDAttributeEmptyValue                = `not creating indexed metric '%s' datapoint: %w`
+	errMsgResourceAttributeEmptyValue           = `related resource attribute value is blank`
+	errMsgOIDResourceAttributeEmptyValue        = `not creating indexed metric '%s' or resource: %w`
+	errMsgScalarOIDProcessing                   = `problem processing scalar metric data for OID '%s': %w`
+	errMsgIndexedMetricOIDProcessing            = `problem processing indexed metric data for OID '%s' from column OID '%s': %w`
+	errMsgIndexedAttributeOIDProcessing         = `problem processing indexed attribute data for OID '%s' from column OID '%s': %w`
+	errMsgIndexedResourceAttributeOIDProcessing = `problem processing indexed resource attribute data for OID '%s' from column OID '%s': %w`
+)
+
 // snmpScraper handles scraping of SNMP metrics
 type snmpScraper struct {
+	client   client
 	logger   *zap.Logger
 	cfg      *Config
 	settings component.ReceiverCreateSettings
 }
 
+// metricsByResourceContainer is a helper alias to hold metric data keyed by resource key and metric name
+type metricsByResourceContainer map[string]map[string]*pmetric.Metric
+
+func newMetricsByResourceContainer() metricsByResourceContainer {
+	mbrc := metricsByResourceContainer{}
+
+	return mbrc
+}
+
+func (m metricsByResourceContainer) getMetrics(resourceKey string) map[string]*pmetric.Metric {
+	if m[resourceKey] == nil {
+		m[resourceKey] = map[string]*pmetric.Metric{}
+	}
+	return m[resourceKey]
+}
+
+// snmpOTELMetricData is a helper data struct to contain OTEL related data that will be used by many functions
+type snmpOTELMetricData struct {
+	// This is used as an easy reference to grab existing OTEL resources by unique key
+	resourcesByKey map[string]*pmetric.ResourceMetrics
+	// The info in this map will ultimately already be contained within the resourcesByKey, but it is
+	// more easily accessible to pull out a specific existing Metric by resource and metric name using this
+	metricsByResource metricsByResourceContainer
+	// This is the ResourceMetricsSlice that will contain all newly created resources and metrics
+	resourceMetricsSlice *pmetric.ResourceMetricsSlice
+}
+
+func newSNMPOTELMetricData(resourceMetricsSlice *pmetric.ResourceMetricsSlice) *snmpOTELMetricData {
+	otelMetricData := snmpOTELMetricData{
+		resourceMetricsSlice: resourceMetricsSlice,
+		resourcesByKey:       map[string]*pmetric.ResourceMetrics{},
+		metricsByResource:    newMetricsByResourceContainer(),
+	}
+
+	return &otelMetricData
+}
```

Comment:
All of this serves the same purpose of providing access to resources & metrics. Let's move it to another file, unify it into one struct and add methods to the struct to interact with it. 

Even better would be to put it in an internal package w/ tests. Resource key generation could live there too, since those keys are required for interacting with the struct.